packages/js/src/cache/notifications-cache.ts

```diff
@@ -0,0 +1,212 @@
+
+      const notificationsResponse = this.getReducedNotifications(filterCache);
+      this.#emitter.emit('notifications.list.updated', {
+        data: notificationsResponse,
+      });
+    });
+  };
+
+  private handleMultipleNotificationsEvent = ({ data }: { data?: Notification[] }): void => {
+    if (!data) {
+      return;
+    }
+
+    const notificationIds = data.map((notification) => notification.id);
+    const filterKeysToUpdate = this.#idsCache.keys().filter((key) => {
+      const idsSet = this.#idsCache.get(key);
+
+      return notificationIds.some((id) => idsSet?.has(id));
+    });
+
+    filterKeysToUpdate.forEach((key) => {
+      const filterCache = this.#cache.get(key);
+      if (!filterCache) {
+        return;
+      }
+
+      data.forEach((notification) => this.updateNotificationInCache(filterCache, notification));
+
+      const notificationsResponse = this.getReducedNotifications(filterCache);
+      this.#emitter.emit('notifications.list.updated', {
+        data: notificationsResponse,
+      });
+    });
+  };
+
+  private has(args: ListNotificationsArgs): boolean {
+    const pageKey = getPageKey(args);
+    const filterCache = this.getFilterCache(args);
+
+    return filterCache?.get(pageKey) !== undefined;
+  }
+
+  private getReducedNotifications(filterCache: Cache<ListNotificationsResponse>): ListNotificationsResponse {
+    const values = filterCache.getValues();
+
+    return values.reduce<ListNotificationsResponse>(
+      (acc, el) => {
+        return {
+          hasMore: el.hasMore,
+          filter: el.filter,
+          notifications: [...acc.notifications, ...el.notifications],
+        };
+      },
+      { hasMore: false, filter: {}, notifications: [] }
+    );
+  }
+
+  set(args: ListNotificationsArgs, data: ListNotificationsResponse): void {
+    const filterKey = getFilterKey(args);
+    const pageKey = getPageKey(args);
+    const filterCache = this.getFilterCache(args) ?? new InMemoryCache();
+    filterCache.set(pageKey, data);
+    this.#cache.set(filterKey, filterCache);
+
+    const uniqueIds = this.#idsCache.get(filterKey) ?? new Set();
+    data.notifications.forEach((el) => uniqueIds.add(el.id));
+    this.#idsCache.set(filterKey, uniqueIds);
+  }
+
+  getAll(args: ListNotificationsArgs): ListNotificationsResponse | undefined {
+    const filterCache = this.getFilterCache(args);
+    if (this.has(args) && filterCache) {
+      return this.getReducedNotifications(filterCache);
+    }
+
+    return;
+  }
+
+  getUniqueNotifications({ tags }: Pick<ListNotificationsArgs, 'tags'>): Array<Notification> {
+    const keys = this.#cache.keys();
+
+    const uniqueNotifications = new Map<string, Notification>();
+    keys.forEach((key) => {
+      const filter: NotificationFilter = JSON.parse(key);
+      if (!tags || tags?.every((tag) => filter.tags?.includes(tag))) {
+        const filterCache = this.#cache.get(key);
+        if (!filterCache) {
+          return;
+        }
```

Comment:
<!-- Suggestions propose improvements to the current subject. It's important to be explicit and clear on what is being suggested and why it is an improvement. Consider using patches and the blocking or non-blocking decorations to further communicate your intent.  -->
✏ **suggestion:** Can you please add a code comment here about what it does.
It seems little confusing why we want to have tags in every filter‏